ccl/sqlproxyccl: replace ConnectionCopy logic in forwarder with interceptors

jaylim-crl opened the pull request:
#### pgwire: expose NewReadTimeoutConn to be used outside of the pgwire package

Previously, readTimeoutConn was only used within the pgwire package to read
messages from the client for the SQL server. We want this construct when
reading messages from clients within the sqlproxy as well. This commit exposes
the readTimeoutConn construct through NewReadTimeoutConn.

#### ccl/sqlproxyccl: replace ConnectionCopy logic in forwarder with interceptors

Previously, we were using io.Copy through ConnectionCopy to forward messages
between the client and SQL server. Now that the interceptors have been merged,
we will update the forwarder to use these interceptors instead of the old
approach.

There are a few notable changes in this commit:
1. We wrap clientConn with the a readTimeoutConn that was exposed in the
  previous commit. This allows us to unblock on Read whenever an activity
  occurs (e.g. context cancellation, and in the future, when transfer is
  requested).
2. There are two goroutines per connection within the forwarder: one for the
  request processor (client to server), and the other for the response
  processor (server to client).
3. We also removed unnecessary checks for codeExpiredClientConnection and
  codeIdleDisconnect errors when copying from server to client. These are
  already handled by the idle monitor's callback as well as the denylist
  watcher's callback. When those constructs were implemented back then, we
  did not remove them from ConnectionCopy.

We need to clean up error messages, and how we propagate them back to the
user because today we just close the connection without returning a response,
resulting in, I believe, a broken pipe error.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/76560)
<!-- Reviewable:end -->

jaylim-crl reviewed (commented):
I'm putting this PR up for review first. Tests will come later once we're aligned on the design. Existing sqlproxyccl tests are all passing - we'll just need to add new ones. As a reminder, only the last two commits are relevant. The first has been reviewed. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76560#-:-MvvToDS6VEW4S-fk1Yl:b-3kvo3p)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/ccl/sqlproxyccl/forwarder.go, line 81 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/76560#-MvvTDtQ1JGEpPmAifMA:-MvvTDtQ1JGEpPmAifMB:b-39bzvm) ([raw file](https://github.com/cockroachdb/cockroach/blob/9b27a48572316127fe6cd5537907d4ff8d449acc/pkg/ccl/sqlproxyccl/forwarder.go#L81)):*
> ```Go
> 		// once the context gets cancelled. This is fine for now since we'll
> 		// be removing this soon anyway.
> 		f.Close()
> ```

This is removed because we no longer need this :)



_Code quote:_
```Go
		// Close the forwarder to clean up. This goroutine is temporarily here
		// because the only way to unblock io.Copy is to close one of the ends,
		// which will be done through closing the forwarder. Once we replace
		// io.Copy with the interceptors, we could use f.ctx directly, and no
		// longer need this goroutine.
		//
		// Note that if f.Close was called externally, this will result
		// in two f.Close calls in total, i.e. one externally, and one here
		// once the context gets cancelled. This is fine for now since we'll
		// be removing this soon anyway.
		f.Close()
```
___
*[pkg/ccl/sqlproxyccl/forwarder.go, line 120 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/76560#-MvvTLgM2YhdZLb_kOHi:-MvvTLgN2iG0OTS7KqRr:b-adpucg) ([raw file](https://github.com/cockroachdb/cockroach/blob/307fcba804fb6ed03062982fe1a2b55d31b09257/pkg/ccl/sqlproxyccl/forwarder.go#L120)):*
> ```Go
> 	if err != nil {
> 		// This will not happen because defaultBufferSize > 5.
> 		panic(errors.NewAssertionErrorWithWrappedErrf(err, "client interceptor"))
> ```

I'd rather we panic here instead of handling the extra error that we won't need in the caller.
___
*[pkg/ccl/sqlproxyccl/forwarder.go, line 204 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/76560#-MvvTjXH7SKoEgY2Jpa4:-MvvTjXH7SKoEgY2Jpa5:b-rnkbm5) ([raw file](https://github.com/cockroachdb/cockroach/blob/307fcba804fb6ed03062982fe1a2b55d31b09257/pkg/ccl/sqlproxyccl/forwarder.go#L204)):*
> ```Go
> }
> 
> // wrapServerToClientError overrides client to server errors for external
> ```

Typo here, but I'll fix it later. Should be "server to client" instead.


<!-- Sent from Reviewable.io -->

jaylim-crl committed:
pgwire: expose NewReadTimeoutConn to be used outside of the pgwire package

Previously, readTimeoutConn was only used within the pgwire package to read
messages from the client for the SQL server. We want this construct when
reading messages from clients within the sqlproxy as well. This commit exposes
the readTimeoutConn construct through NewReadTimeoutConn.

Release note: None

jaylim-crl committed:
ccl/sqlproxyccl: replace ConnectionCopy logic in forwarder with interceptors

Informs #76000.

Previously, we were using io.Copy through ConnectionCopy to forward messages
between the client and SQL server. Now that the interceptors have been merged,
we will update the forwarder to use these interceptors instead of the old
approach.

There are a few notable changes in this commit:
1. We wrap clientConn with the a readTimeoutConn that was exposed in the
   previous commit. This allows us to unblock on Read whenever an activity
   occurs (e.g. context cancellation, and in the future, when transfer is
   requested).
2. There are two goroutines per connection within the forwarder: one for the
   request processor (client to server), and the other for the response
   processor (server to client).
3. We also removed unnecessary checks for codeExpiredClientConnection and
   codeIdleDisconnect errors when copying from server to client. These are
   already handled by the idle monitor's callback as well as the denylist
   watcher's callback. When those constructs were implemented back then, we
   did not remove them from ConnectionCopy.

We need to clean up error messages, and how we propagate them back to the
user because today we just close the connection without returning a response,
resulting in, I believe, a broken pipe error.

Release note: None

jaylim-crl reviewed (commented):
This PR is ready for a review. Only review the last two commits. The others are currently in the merge queue; they have been in there for hours and bors is unhappy. Once they are in master, I'll rebase.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76560#-:-Mw-rkPe-FibpYUX6-z-:b-eeinwa)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andy-kimball and @JeffSwenson)

___
*[pkg/ccl/sqlproxyccl/forwarder.go, line 120 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/76560#-MvvTLgM2YhdZLb_kOHi:-Mw-rf-pEF7LT_ILtLwo:bav9om4) ([raw file](https://github.com/cockroachdb/cockroach/blob/307fcba804fb6ed03062982fe1a2b55d31b09257/pkg/ccl/sqlproxyccl/forwarder.go#L120)):*
<details><summary><i>Previously, jaylim-crl (Jay Lim) wrote…</i></summary><blockquote>

I'd rather we panic here instead of handling the extra error that we won't need in the caller.
</blockquote></details>

No longer needed.
___
*[pkg/ccl/sqlproxyccl/forwarder.go, line 204 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/76560#-MvvTjXH7SKoEgY2Jpa4:-Mw-rhsr0nRiGJJhq2LF:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/307fcba804fb6ed03062982fe1a2b55d31b09257/pkg/ccl/sqlproxyccl/forwarder.go#L204)):*
<details><summary><i>Previously, jaylim-crl (Jay Lim) wrote…</i></summary><blockquote>

Typo here, but I'll fix it later. Should be "server to client" instead.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

JeffSwenson commented on pkg/ccl/sqlproxyccl/forwarder.go:
```diff
@@ -105,3 +129,78 @@ func (f *forwarder) Close() {
+// unfortunately. At the moment, this is only used for metrics. See TODO in
+// proxy_handler about sending safely to avoid corrupted packets. Handle these
+// errors in a friendly manner.
+func (f *forwarder) wrapClientToServerError(err error) error {
```
nit: wrapClientToServerError and wrapServerToClientError can be implemented as functions instead of methods.

jeffswenson reviewed (approved):
LGTM

jaylim-crl reviewed (commented):
TFTR!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76560#-:-Mw2y-pl2sOXXvmn3okT:b-814g8k)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andy-kimball)

___
*[pkg/ccl/sqlproxyccl/forwarder.go, line 154 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/76560#-MvvT8j_rO2ygmJqGP9W-r5-154:-Mw2y136Awc0gz5ejyt-:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/bb437f9ec819e615f5e6eaaaa496929dc66c3e73/pkg/ccl/sqlproxyccl/forwarder.go#L154)):*
<details><summary><i>Previously, JeffSwenson (Jeff Swenson) wrote…</i></summary><blockquote>

nit: wrapClientToServerError and wrapServerToClientError can be implemented as functions instead of methods.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

jaylim-crl commented:
bors r=JeffSwenson
